### ISSUE TITLE
NFSv3 and v4 shares

### ISSUE THREAD

**Issue by sbyv** [derdanne/puppet-nfs#59]
Hi Daniel,

This is more of a question rather than an issue really.

Is there a way to get your nfs module to set up NFS v3 and v4 shares/exports on a single NFS server? So far the only way to achieve that I found was a workaround like this:

```
  nfs::server::export { '/home':
    ensure  => present,
    clients => '*(sec=krb5:krb5i:krb5p,rw,blah_blah_blah)',
  }
  file_line { '/home NFS 3':
    path  => '/etc/exports',
    line => '/home 10.0.0.0/24(rw,blah_blah_blah),
    notify  => Service[$::nfs::server_service_name],
  }
```
that works but is far from ideal as with every puppet agent run, your nfs module overwrites /etc/exports, removing all file_line entries and right after that file_line resources modify it adding missing entries again.

Any hints much appreciated!

Cheers,
Adam

**Comment by derdanne** [comment]
Hi Adam,

the NFS version should't be considered on the "exports" level. The NFS Server Daemon itself can export all resources for different NFS client versions. You can achieve the backwards compatibility by configuring  options for mountd. 

In example on a Gentoo host we have to set the following options to be compatible to NFS versions 2, 3 and 4 (like described on http://nfs.sourceforge.net in section A4):

# Options to pass to rpc.mountd
OPTS_RPC_MOUNTD="-V 4 -V 3 -V 2"

Since we currently do not yet support this setting within this module, you have to set it on your own.

Cheers back,
Daniel

**Comment by sbyv** [comment]
Thanks for your immediate reply, Daniel!

Okay, so having NFS server configured like this:
```
  class { '::nfs':
    server_enabled  => true,
    nfs_v4          => true,
    nfs_v4_idmap_domain => 'my_domain',
    nfs_v4_export_root  => '/shared',
    nfs_v4_export_root_clients => '*(sec=krb5:krb5i:krb5p,rw,sync,insecure,wdelay,no_root_squash,no_subtree_check,fsid=0)',
  }
```
and a share:

**Comment by derdanne** [comment]
Hi,

when you set the option for MOUNTD as described above, the selection of the version is considered on the client side, you do not have to specify the version on the server for the export.

When you use this module also on your NFS clients, you should set "nfs_v4_client => false".

You can test the mount with a specific version on a client by issuing (i.e version 3, following your example):

mount -t nfs -o nfsvers=3 nfsserver.example.com:/shared/trash /trash

Cheers
Daniel

**Comment by sbyv** [comment]
Sorry mate, we have a bit of a misunderstanding here, let me rephrase and clarify.

I do have NFS v3 and v4 enabled on the server, that's sorted. I can mount NFS shares via v3 and v4 on a client off the server, problem is, v3 only works with my file_line workaround I mentioned in the first post. If I remove it, /files/trash gets removed as well and only /shared/trash persists, as per nfs::server::export { '/files/trash': clients => '*(sec=krb5:krb5i:krb5p...)' } resource, but that's only for Kerberized, NFS v4 clients. If I attempt to mount it via NFS v3:

**Comment by derdanne** [comment]
Hi,

sorry, this module will bind-mount your "/files/trash" to the "$nfs_v4_export_root/trash". 

You can't have both of them in your exports with the current module, since the exported directory  depends on "$nfs::server::nfs_v4", which you can't override on the exports level (compare https://github.com/derdanne/puppet-nfs/blob/master/manifests/server/export.pp#L95). 

But i do not understand, why you need to have the same data available in another location. Could you give me a hint, why someone would need this? Maybe we can submit a feature request for this scenario.

Cheers,
Daniel

**Comment by sbyv** [comment]
Hi,

Yes, as per my second comment - I have noticed the automatic bind-mount ;)
As for not being able to have both exports, I thought as much yet decided to ask in case I'd overlooked something.

The only reason I needed that feature was to be able to support some of my legacy production clients, and there's plenty of those, with fs:/files/... nfs mounts and they're not managed by puppet and I've been trying to avoid as much as possible making changes to the production environment, which is not a trivial task in my company.

Thanks for your time and explanations, I guess I'll continue with my ugly workaround until I'm allowed to change mounts on the legacy clients. Unless you'd be interested in my contribution to the module, I might try to implement that bit. Not that there is an urgent need for that, it's just for the sake of flexibility to allow others to use the module in a bit unusual scenario like mine, if needed.

Cheers,
Adam

**Comment by derdanne** [comment]
Hi,

ok, now i completely understood your scenario. Feel free to add this feature via pull request, i really appreciate any contribution to this module.

Cheers
Daniel
